sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,535 @@
+static int bcmemmc_set_clk_freq(device_t *dev, uint32_t frq) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  int cnt;
+
+  /* TODO(mohrcore): Not sure if this is necessary. If the will run fine on
+   * a real hardware without it, it should be removed. */
+  cnt = NSPINS;
+  while (b_in(emmc, BCMEMMC_STATUS) & SR_INHIBIT) {
+    if (--cnt < 0)
+      return ETIMEDOUT;
+    delay(3);
+  }
+
+  b_clr(emmc, BCMEMMC_CONTROL1, C1_CLK_EN);
+  /* host_version <= HOST_SPEC_V2 needs a power-of-two divisor.
+   * It would require a different calculation method. */
+  assert(state->host_version > HOST_SPEC_V2);
+  bcmemmc_clk_set_divisor(state, frq);
+  b_set(emmc, BCMEMMC_CONTROL1, C1_CLK_EN);
+
+  /* Wait until the clock becomes stable */
+  cnt = NSPINS;
+  while (!(b_in(emmc, BCMEMMC_CONTROL1) & C1_CLK_STABLE)) {
+    if (--cnt < 0)
+      return ETIMEDOUT;
+    delay(30);
+  }
+
+  return 0;
+}
+
+/* Encode `cmd` to command an appropriate command register value */
+static uint32_t bcmemmc_encode_cmd(emmc_cmd_t cmd) {
+  uint32_t code = (uint32_t)cmd.cmd_idx << 24;
+
+  switch (cmd.exp_resp) {
+    case EMMCRESP_NONE:
+      break;
+    case EMMCRESP_R2:
+      code |= CMD_RESP136;
+      break;
+    case EMMCRESP_R1B:
+      code |= CMD_RESP48B;
+      break;
+    default:
+      code |= CMD_RESP48;
+  }
+  switch (emmc_cmdtype(cmd.flags)) {
+    case EMMC_F_SUSPEND:
+      code |= CMD_TYPE_SUSPEND;
+      break;
+    case EMMC_F_RESUME:
+      code |= CMD_TYPE_RESUME;
+      break;
+    case EMMC_F_ABORT:
+      code |= CMD_TYPE_ABORT;
+      break;
+    default:
+      break;
+  }
+  if (cmd.flags & EMMC_F_DATA_READ)
+    code |= CMD_DATA_TRANSFER | CMD_DATA_READ;
+  if (cmd.flags & EMMC_F_DATA_WRITE)
+    code |= CMD_DATA_TRANSFER;
+  if (cmd.flags & EMMC_F_DATA_MULTI)
+    code |= CMD_DATA_MULTI;
+  if (cmd.flags & EMMC_F_CHKIDX)
+    code |= CMD_CHECKIDX;
+  if (cmd.flags & EMMC_F_CHKCRC)
+    code |= CMD_CHECKCRC;
+
+  return code;
+}
+
+static int bcmemmc_set_bus_width(bcmemmc_state_t *state, uint32_t bw) {
+  switch (bw) {
+    case EMMC_BUSWIDTH_1:
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_8BIT);
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_DWITDH);
+      return 0;
+    case EMMC_BUSWIDTH_4:
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_8BIT);
+      b_set(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_DWITDH);
+      return 0;
+    case EMMC_BUSWIDTH_8:
+      b_set(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_8BIT);
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_DWITDH);
+      return 0;
+    default:
+      return EINVAL;
+  }
+}
+
+static int bcmemmc_get_bus_width(bcmemmc_state_t *state) {
+  uint32_t ctl = b_in(state->emmc, BCMEMMC_CONTROL0);
+  if (ctl & C0_HCTL_8BIT)
+    return EMMC_BUSWIDTH_8;
+  if (ctl & C0_HCTL_DWITDH)
+    return EMMC_BUSWIDTH_4;
+  return EMMC_BUSWIDTH_1;
+}
+
+#define MAXBLKSIZE 512 /* Currently only blocks of 512 bytes are supported. */
```

Comment:
According to `bcmemmc_set_prop`:
 - `MAXBLKSIZE` should be 1024 (should it always be a power of two bigger than 0?)
 - `MAXBLKCNT` should be 65535 (cannot be zero – right?)

Reply:
How does a 10-bit value imply `MAXBLKSIZE = 1024`? It's 1023 at most and no one is going to use that. I'm not even sure if non-power-of-two block sizes are supported by that controller.

I've bumped the value of `MAXBLKCNT` to 65535 tho.